src/Domain/JapaneseParser.php

```diff
@@ -97,4 +97,29 @@ public function getParsedTokens(string $text, Language $lang) {
         return $tokens;
     }
 
+    public function getReading(string $text, Language $lang) {
+        // Japanese is an exception
+        $mecab_file = tempnam(sys_get_temp_dir(), "mecab_to_db");
+        $mecab_args = ' -O yomi ';
+        if (file_exists($mecab_file)) { 
+            unlink($mecab_file); 
+        }
+        $fp = fopen($mecab_file, 'w');
+        fwrite($fp, $text . "\n");
+        fclose($fp);
+        $mecab = JapaneseParser::MeCab_command($mecab_args);
+        $handle = popen($mecab . $mecab_file, "r");
+        // $mecab_str Output string
+        $mecab_str = '';
+        while (($line = fgets($handle, 4096)) !== false) {
+            $mecab_str .= $line; 
+        }
+        if (!feof($handle)) {
+            echo "Error: unexpected fgets() fail\n";
```

Comment:
Code should never `echo` or `dump` or `die`, it interrupts flow.  It should `throw new \Exception(` .  But I wonder what the right thing to do would be, to really throw, or just to return a blank string?

Reply:
My bad, I copy/pasted a bit too fast. If the stream $handle is not closed yet, it means the file is unreadable. I'm not sure how you want to fix it:
- If it doesn't work, the MeCab parser won't work: maybe a ``die`` would be necessary?
- It's a not critical part of code: a pronunciation left blank + a console message should be enough.

I would go for the latter, as it means less interruptions for the user.

Reply:
"die" interrupts everything and bypasses any top-level error handling, so instead code really should throw so that the exception can be caught at a higher level. ... BUT in this case, if mecab's not working for getting readings, then killing the whole program is going to be overkill. :-)

I agree the console thing is best, but it might be tough to implement, actually!  The code to handle the reading happens deep in the app code, but the messaging for to the console is currently only managed in the controller layer (where it should be).

Given that consideration, while it's not the greatest feeling, how about you just ... silently catch and ignore any exception, and return an empty string?  Add a comment explaining why the error is being ignored.  It's not great, but who cares?  If the readings don't get added, the user hasn't _lost_ anything, and if they work, great.  Does that sound reasonable?

Reply:
Code successfully fails silently!